--update-all and --interactive

peterbe opened the pull request:
Fixes #90

peterbe commented:
@mythmon I don't need an in-depth review but can you take a look?

Basically, the way I test it is:

```bash
python hashin.py -r ~/songsearch/requirements.txt -i -u --dry-run
```

Do you like the interactive prompt? Is it working? Anything else obvious you see?

mythmon commented on hashin.py:
```diff
@@ -196,6 +158,30 @@ def run_packages(
+                    first_interactive = False
+                    continue
+                first_interactive = False
+            except InteractiveAll:
```
My general rule is to only use exceptions for exceptional cases. "All" and "Quit" don't seem exceptional to me. I think I would make `interactive_upgrade_request` return a set of constants for all the cases, instead of returning a boolean in some cases and raising exception in others.

peterbe replied:
Yeah, I tend to agree. ...in hindsight. I think I fell into the trap of doing it like this because I had originally not planned to use "all" and "quit" and when I added those it didn't match the way I called the `interactive_upgrade_request` function.

mythmon commented on hashin.py:
```diff
@@ -228,6 +219,69 @@ def run_packages(
+            clear_line()
+            print_line(True)
+            return True
+        return ask()
```
It might be better to put this in a loop instead of recursing. I think that makes the repetition easier to read.

peterbe replied:
Point taken. I can see how that could be used to achieve the same thing. But I kinda like it as a function like that. Perhaps it fits my brain better.

mythmon commented on hashin.py:
```diff
@@ -228,6 +219,69 @@ def run_packages(
+        sys.stdout.write("\033[K")  # Clear to the end of line
+
+    def ask():
+        answer = input("Update? [Y/n/a/q]: ").lower().strip()
```
As a nifty improvement, it would be cool to add "?" to the list of options, and if ? or an invalid choice is given, it prints a longer version of this prompt out, like:

```
y - Include this upgrade (default)
n - Skip this upgrade
a - Include this and all following upgrades
q - Skip this and all following upgrades
? - Print this help
```

You can see a similar UI in `git add -i`, which prompts you to include, skip, edit, etc chunks of files.

peterbe replied:
+1

peterbe commented:
@mythmon Wanna take a last look? The recent changes are quite insignificant. In particular, I discovered a bug in my first attempt where if you had `some-package=1.2.3` in your requirements file but, over time, that package has been renamed (301 in Pypi JSON lingo) to `other-package` then my it wouldn't work. Now it works and there are unit tests for that too. 

The other thing I changed was that I add the "?" option to the prompt. Thank you. Also, I changed it so that if you do say "A" ("all") it will "keep the interactive", but not ask any more questions. That way you can clearly see which packages can be updated and what they get updated to.
